### PR TITLE
Add bqStrictSchemaDocTypes option

### LINKED ISSUE #644: Add option to not fill `additionalProperties` for specific schemas/docTypes
Specifically, we are looking at the main ping. To use table splitting, each table will be "missing" parts of the schema. We want these dropped instead of filled.

Short term, we can have a small main-ping schema, which includes the environment and common ping sections.  We would prefer this be used without additionalProperties as well, to reduce storage cost.

### PR DESCRIPTION
**Pull request by jklukas**
Closes #644

We should be able to set the following for the bq sink on both telemetry
and structured:

    bqStrictSchemaDocTypes=main

### CODE REVIEW

**Comment by jklukas** [comment]
One detail here is that I'm choosing to still remove fields from the payload if they don't exist in the BQ schema. That should lead to less data to schlep to BigQuery. We could instead choose to leave them in for docTypes we've opted into strict schema handling, and then let BQ ignore them.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java [hunk 1]
```diff
@@ -383,8 +387,8 @@ public BigQueryOutput(ValueProvider<String> tableSpecTemplate, BigQueryWriteMeth
 
       streamingInput.ifPresent(messages -> {
         WriteResult writeResult = messages //
-            .apply(PubsubMessageToTableRow.of(tableSpecTemplate)).errorsTo(errorCollections)
-            .apply(baseWriteTransform //
+            .apply(PubsubMessageToTableRow.of(tableSpecTemplate, strictSchemaDocTypes))
+            .errorsTo(errorCollections).apply(baseWriteTransform //
```
Comment (fbertsch) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java]: nit: should `errorsTo` be indented, and the following `apply` be uninedented?

File: ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java [hunk 2]
```diff
@@ -167,12 +172,24 @@ private PubsubMessageToTableRow(ValueProvider<String> tableSpecTemplate) {
     // Strip metadata so that it's not subject to transformation.
     Object metadata = tableRow.remove(AddMetadata.METADATA);
 
+    String namespace = message.getAttribute(ParseUri.DOCUMENT_NAMESPACE);
+    String docType = message.getAttribute(ParseUri.DOCUMENT_TYPE);
+    final boolean strictSchema;
+    if (("telemetry".equals(namespace) && strictSchemaDocTypes.get().contains(docType))
```
Comment (fbertsch) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java]: This special-casing for telemetry seems unnecessarily, we could remove it (which would simplify the documentation above and the implementation here).

Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java]: Perhaps in retrospect that would have been a good idea. We already support this format for other options that take in a comma separated list of docTypes, though, so I'm inclined to keep consistent.

**Comment by jklukas** [comment]
> Makes sense for preserving existing behavior. We could consider removing as much special-casing of telemetry behavior as we can at some point in the future.

I have become sufficiently convinced that this is a good idea and that this will be easier to change now than later that I added in the change to this PR. All options that accept lists of docTypes now expect entries to be qualified with a namespace.

So the bqSink configuration will need to be:

    bqStrictSchemaDocTypes=telemetry/main

And the Republisher configuration will need to change the following:

    perDocTypeEnabledList=telemetry/event,telemetry/prio

No change is necessary for `bqStreamingDocTypes` since the values there are already fully qualified.

I will not merge this until I can confirm with @whd that we're ready with the updated configuration.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java (hunk 2 above)
Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java]: You're right. I've changed this.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java [hunk 3]
```diff
@@ -167,12 +172,20 @@ private PubsubMessageToTableRow(ValueProvider<String> tableSpecTemplate) {
     // Strip metadata so that it's not subject to transformation.
     Object metadata = tableRow.remove(AddMetadata.METADATA);
 
+    String namespace = message.getAttribute(ParseUri.DOCUMENT_NAMESPACE);
+    String docType = message.getAttribute(ParseUri.DOCUMENT_TYPE);
+    final boolean strictSchema = (strictSchemaDocTypes.isAccessible()
+        && (("telemetry".equals(namespace) && strictSchemaDocTypes.get().contains(docType))
```
Comment (fbertsch) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java]: Can't this special-casing be removed now?

File: ingestion-beam/src/test/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRowTest.java [hunk 4]
```diff
@@ -123,6 +123,19 @@ public void testAdditionalProperties() throws Exception {
     assertEquals(expectedAdditional, Json.asString(additionalProperties));
   }
 
+  @Test
+  public void testStrictSchema() throws Exception {
+    Map<String, Object> parent = new HashMap<>();
+    parent.put("outer", new HashMap<>(ImmutableMap.of("otherfield", 3, //
```
Comment (fbertsch) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRowTest.java]: Might be good to add a telemetry test now.

**Comment by whd** [comment]
I've updated the template builds to include fully qualified versions of the config options.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRowTest.java (hunk 4 above)
Comment (jklukas) [reply ingestion-beam/src/test/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRowTest.java]: Added an integration test that exercises `bqStrictSchemaDocTypes`

**Pull request merged**